File: python/binary-search.py
class BinarySearch():
    
  def search_iterative(self, array, targetNumber):
    low = 0
    high = len(list) - 1
    count = 0

    while low <= high:
      middle = (low + high) // 2
      guess = list[middle]

      if guess == targetNumber:
        print('steps to find: ', count)
        return middle
      if targetNumber < guess:
        count += 1
        high = middle - 1
      else:
        count += 1
        low = middle + 1
    return -1
  
if __name__ == "__main__":
  bs = BinarySearch()
  my_list = [1,3,5,7,9]

  index = bs.search_iterative(my_list, 3) # => 1

  if index != -1:
    print('Number found at index', index)
  else:  
    print('Number not found at index')
    

File: javascript/binarySearch.js
/**
 * Running time = O(log n)
 * Ω(1) in best case
 * Binary Search in a array of numbers ordered
 * @param {number} search number in array 
 * @param {Array} array of numbers ordered 
 * @returns {number} index in array where number was found
 */
function binarySearch(search, arr) {
  let low = 0
  let high = arr.length - 1
  while (low <= high) {
    let middle = Math.floor((low + high) / 2)
    let guess = arr[middle]
    if (guess === search) {
      return middle
    } else if (guess < search) {
      low = middle + 1
    } else {
      high = middle - 1
    }
  }
  return -1
}

const arr = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];
const index = binarySearch(13, arr);

if (index != -1) {
  console.log(`Number found at index %d`, index)
} else {
  console.log(`Number not found at index`)
}